src/FsAutoComplete.Core/InlayHints.fs

```diff
@@ -313,46 +313,521 @@ module private ShouldCreate =
+            let typed = isDirectlyTyped ident.idRange.Start path
+
+            if typed then
+              ExplicitType.Exists |> Some
+            else
+              let parens = getParensForIdentPat text pat path
+
+              ExplicitType.Missing
+                { Ident = ident.idRange
+                  InsertAt = ident.idRange.End
+                  Parens = parens
+                  SpecialRules = [ RemoveOptionFromType ]
+                //              ^^^^^^^^^^^^^^^^^^^^
+                //              `?v: int`, NOT `?v: int option`
+                }
+              |> Some
+          | _ -> defaultTraverse pat //todo: custom traverse? -> doesn't require FCS to handle `SynPat.Record`
+
+        member _.VisitSimplePats(path, pats) =
+          // SynSimplePats at:
+          // * Primary ctor:
+          //    * SynMemberDefn.ImplicitCtor.ctorArgs
+          //    * SynTypeDefnSimpleRepr.General.implicitCtorSynPats
+          //      //TODO: when? example?
+          // * Lambda: SynExpr.Lambda.args
+          //   * issue: might or might not be actual identifier
+          //      * `let f1 = fun v -> v + 1`
+          //         -> `v` is in `args` (-> SynSimplePat)
+          //      * `let f2 = fun (Value v) -> v + 1`
+          //        -> compiler generated `_arg1` in `args`,
+          //           and `v` is inside match expression in `body` & `parsedData` (-> `SynPat` )
+          maybe {
+            let! pat =
+              pats
+              |> List.tryFind (fun p -> rangeContainsPos p.Range pos)
+
+            let rec tryGetIdent pat =
+              match pat with
+              | SynSimplePat.Id (ident = ident) when rangeContainsPos ident.idRange pos -> Some pat
+              | SynSimplePat.Attrib (pat = pat) when rangeContainsPos pat.Range pos -> tryGetIdent pat
+              | SynSimplePat.Typed (pat = pat) when rangeContainsPos pat.Range pos -> tryGetIdent pat
+              | _ -> None
+
+            let! ident = tryGetIdent pat
+
+            match ident with
+            | SynSimplePat.Id(isCompilerGenerated = false) ->
+              let rec isTyped =
+                function
+                | SynSimplePat.Typed _ -> true
+                | SynSimplePat.Id _ -> false
+                | SynSimplePat.Attrib (pat = pat) -> isTyped pat
+
+              let typed = isTyped pat
+
+              if typed then
+                return ExplicitType.Exists
+              else
+                let isCtor =
+                  path
+                  |> List.tryHead
+                  |> Option.map (function
+                    // normal ctor in type: `type A(v) = ...`
+                    | SyntaxNode.SynMemberDefn (SynMemberDefn.ImplicitCtor _) -> true
+                    //TODO: when? example?
+                    | SyntaxNode.SynTypeDefn (SynTypeDefn(typeRepr = SynTypeDefnRepr.Simple(simpleRepr = SynTypeDefnSimpleRepr.General(implicitCtorSynPats = Some (ctorPats))))) when
```

Comment:
When does this happen?
Case above is normal primary ctor (`type A(v) = ...`) -- but when does this here occur?

Reply:
Unknown to me 🤷